pkg/mq/msgdispatcher/manager.go

```diff
@@ -67,105 +67,78 @@ func NewDispatcherManager(pchannel string, role string, nodeID int64, factory ms
-		return nil, fmt.Errorf("solo vchannel dispatcher already exists")
-	}
-	if c.mainDispatcher != nil {
-		if _, err := c.mainDispatcher.GetTarget(vchannel); err == nil {
-			// current dispatcher didn't allow multiple subscriptions on same vchannel at same time
-			log.Warn("unreachable: vchannel has been registered in main dispatcher, ")
-			return nil, fmt.Errorf("vchannel has been registered in main dispatcher")
-		}
-	}
-
-	isMain := c.mainDispatcher == nil
-	d, err := NewDispatcher(ctx, c.factory, isMain, c.pchannel, streamConfig.Pos, c.constructSubName(vchannel, isMain), streamConfig.SubPos, c.lagNotifyChan, c.lagTargets, false)
-	if err != nil {
-		return nil, err
-	}
-	t := newTarget(vchannel, streamConfig.Pos, streamConfig.ReplicateConfig)
-	d.AddTarget(t)
-	if isMain {
-		c.mainDispatcher = d
-		log.Info("add main dispatcher")
-	} else {
-		c.soloDispatchers[vchannel] = d
-		log.Info("add solo dispatcher")
+	t := newTarget(streamConfig)
+	if _, ok := c.registeredTargets.GetOrInsert(t.vchannel, t); ok {
+		return nil, fmt.Errorf("vchannel %s already exists in the dispatcher", t.vchannel)
 	}
-	d.Handle(start)
+	log.Info("target register done", zap.String("vchannel", t.vchannel))
 	return t.ch, nil
 }
 
 func (c *dispatcherManager) Remove(vchannel string) {
-	log := log.With(zap.String("role", c.role),
-		zap.Int64("nodeID", c.nodeID), zap.String("vchannel", vchannel))
+	log := log.With(zap.String("role", c.role), zap.Int64("nodeID", c.nodeID), zap.String("vchannel", vchannel))
+
+	t, ok := c.registeredTargets.GetAndRemove(vchannel)
+	if !ok {
+		log.Info("the target was not registered before", zap.String("vchannel", vchannel))
+		return
+	}
 	c.mu.Lock()
 	defer c.mu.Unlock()
-	if _, ok := c.soloDispatchers[vchannel]; ok {
-		c.soloDispatchers[vchannel].Handle(terminate)
-		c.soloDispatchers[vchannel].CloseTarget(vchannel)
-		delete(c.soloDispatchers, vchannel)
-		c.deleteMetric(vchannel)
-		log.Info("remove soloDispatcher done")
-	}
-	c.lagTargets.GetAndRemove(vchannel)
 
-	if c.mainDispatcher != nil {
+	for _, dispatcher := range c.deputyDispatchers {
+		if dispatcher.HasTarget(vchannel) {
+			dispatcher.Handle(pause)
+			dispatcher.RemoveTarget(vchannel)
+			if dispatcher.TargetNum() == 0 {
+				dispatcher.Handle(terminate)
+				delete(c.deputyDispatchers, dispatcher.ID())
+				log.Info("remove deputy dispatcher done", zap.Int64("id", dispatcher.ID()))
+			} else {
+				dispatcher.Handle(resume)
+			}
+		}
+	}
+	if c.mainDispatcher != nil && c.mainDispatcher.HasTarget(vchannel) {
 		c.mainDispatcher.Handle(pause)
-		c.mainDispatcher.CloseTarget(vchannel)
-		if c.mainDispatcher.TargetNum() == 0 && len(c.soloDispatchers) == 0 {
+		c.mainDispatcher.RemoveTarget(vchannel)
+		if c.mainDispatcher.TargetNum() == 0 && len(c.deputyDispatchers) == 0 {
 			c.mainDispatcher.Handle(terminate)
 			c.mainDispatcher = nil
-			log.Info("remove mainDispatcher done")
 		} else {
 			c.mainDispatcher.Handle(resume)
 		}
 	}
+
+	t.close()
```

Comment:
why close?

Reply:
Because we need to trigger closing flowgraph.